meson.build

```diff
@@ -0,0 +1,362 @@
+else
+  cdata.set_quoted('DBUSPROXY', get_option('system_dbus_proxy'))
+endif
+
+# Flatpak is Linux-specific, so for now we assume that -fvisibility=hidden
+# is always supported
+cdata.set('FLATPAK_EXTERN', '__attribute__((visibility("default"))) extern')
+
+if glib_dep.version().version_compare('>=2.60')
+  # Ignore massive GTimeVal deprecation warnings in 2.62
+  cdata.set('GLIB_VERSION_MIN_REQUIRED', 'GLIB_VERSION_2_60')
+endif
+
+if appstream_dep.version().version_compare('>=0.14.0')
+  cdata.set('HAVE_APPSTREAM_0_14_0', 1)
+endif
+
+if libseccomp_dep.found()
+  cdata.set('ENABLE_SECCOMP', 1)
+endif
+
+if xau_dep.found()
+  cdata.set('ENABLE_XAUTH', 1)
+endif
+
+if not get_option('sandboxed_triggers')
+  cdata.set('DISABLE_SANDBOXED_TRIGGERS', 1)
+endif
+
+if cc.has_function(
+  'archive_read_support_filter_all',
+  dependencies : libarchive_dep,
+  prefix : '#include <archive.h>',
+)
+  cdata.set('HAVE_ARCHIVE_READ_SUPPORT_FILTER_ALL', 1)
+endif
+
+configure_file(
+  output : 'config.h',
+  configuration : cdata,
+)
+
+if meson.version().version_compare('>=0.53.0')
+  summary({
+    'Build system helper' : build_system_helper,
+    'Build selinux module' : build_selinux_module,
+    'Build bubblewrap' : (get_option('system_bubblewrap') == ''),
+    'Build dbus-proxy' : (get_option('system_dbus_proxy') == ''),
+    'Use sandboxed triggers' : get_option('sandboxed_triggers'),
+    'Use seccomp' : libseccomp_dep.found(),
+    'Privileged group' : get_option('privileged_group'),
+    'Use dconf' : dconf_dep.found(),
+    'Use libsystemd' : libsystemd_dep.found(),
+    'Use libmalcontent' : malcontent_dep.found(),
+    'Use libzstd' : libzstd_dep.found(),
+    'Use auto sideloading' : get_option('auto_sideloading'),
+  })
+endif
+
+if get_option('system_bubblewrap') == ''
+  subproject('bubblewrap', default_options : ['program_prefix=flatpak-'])
+endif
+
+if get_option('system_dbus_proxy') == ''
+  subproject('dbus-proxy', default_options : ['program_prefix=flatpak-'])
+endif
+
+# Used for .service files in multiple subdirectories
+service_conf_data = configuration_data()
+service_conf_data.set('libexecdir', get_option('prefix') / get_option('libexecdir'))
+service_conf_data.set('localstatedir', get_option('prefix') / get_option('localstatedir'))
+service_conf_data.set('media_dir', get_option('prefix') / get_option('run_media_dir'))
+service_conf_data.set('extraargs', '')
+
+subdir('common')
+subdir('data')
+
+subdir('app')
+subdir('icon-validator')
+subdir('oci-authenticator')
+subdir('portal')
+subdir('revokefs')
+subdir('session-helper')
+
+subdir('completion')
+subdir('doc')
+subdir('po')
+subdir('tests')
+subdir('triggers')
+
+if get_option('auto_sideloading')
+  subdir('sideload-repos-systemd')
+endif
+
+if build_selinux_module
+  subdir('selinux')
+endif
+
+if build_system_helper
+  subdir('system-helper')
+endif
+
+# TODO: needs 0.62.0: {'pkgconfig.generate implicit variable for builtin directories'}
```

Comment:
For backwards-compat with older versions you can just define datadir yourself as

```meson
'datadir=@0@'.format('${prefix}' / get_option('datadir'))
```

It's a bit ugly, which is why I made this automatic in newer versions of Meson... but that's not too bad for backwards-compat.

Reply:
Thanks for the tip. I need to actually test this on 0.49 to confirm that I've done everything in a genuinely backwards-compatible way (at the moment it compiles on Debian unstable, I haven't tried running tests, and I haven't tried older Meson).